ibis/expr/analysis.py

```diff
@@ -1207,3 +1207,55 @@ def _visit_table(self, expr):
                 and base_table.op() not in self.query_roots
             ):
                 self.foreign_table = expr
+
+
+@functools.singledispatch
```

Comment:
What do you think about creating a new package just for rewrite rules (one file per purpose)?

Each file should have a well defined behaviour on the input expression.

Reply:
Yeah, I think that makes sense. Let's do it in a follow up.